Allow a default handler to be used

duncan3dc opened the pull request:
I want to take advantage of the content type guessing this library does, while at the same time specifying which editor the `PrettyPageHandler` should use.

This PR allows this behaviour by supporting a default handler to be used if no content specific one is in use

oscarotero commented:
Good idea 👍 
I prefer it to be configured using a method, instead a constructor. Do you mind to change it to something like this?

```php
$middleware[] = (new Whoops())
    ->defaultHandler(new PlainTextHandler());
```

And if you can provide a test and edit the README with this new option, it would be great
Thanks!

duncan3dc commented:
A method sounds fine to me.

I had a look for tests but couldn't find any for that section of code.

It looks like the content guessing is skipped when running tests due to a check for cli mode.

How would you like the tests to handle this?

oscarotero commented:
Mmm, I see.
I think this can be tested mocking the `php_sapi_name` native function.
You can see an example here: https://github.com/middlewares/client-ip/blob/master/tests/ClientIpTest.php#L136
this middleware uses the native function `file_get_contents` that is mocked in the tests using the `eloquent/phony-phpunit` package.

duncan3dc commented:
I've pushed some general tests for the content guessing code now, and added a test for the new default handler, I've also moved it into a method rather than the constructor.

Taking advantage of the root scope fallback in PHP doesn't seem like a good idea to me, especially as it's common to explicitly import root functions so that opcache can inline them away in some cases.
So instead of doing that, I've used the [uopz extension](http://php.net/manual/en/book.uopz.php) to mock the call to `php_sapi_name()`

oscarotero commented:
I understand.
The problem is that it requires to install the uopz extension to run the tests and i'd prefer to avoid that if with Phony we have the same result with more compatibility and easy to install.

duncan3dc commented:
I can skip the tests if `uopz` isn't installed, and make it an optional dependency if you'd like?

oscarotero commented:
> I understand.

Well, I lied :)
I don't understand what is the problem with Phony to do not use it for testing? I've tried this and works fine:

```php
public function setUp()
    {
        Phony::stubGlobal('php_sapi_name', 'Middlewares')->returns('phpunit');
    }

    public function tearDown()
    {
        Phony::restoreGlobalFunctions();
    }
```

duncan3dc commented:
It works at the moment because you're using `php_sapi_name()` in the `Middlewares` namespace.
Because a function called `php_sapi_name()` doesn't exist in the `Middlewares` namespace, PHP then goes and looks for a function in the root namespace (`\php_sapi_name()`).

This is a double function lookup, so is wasted effort, and it is currently one of the big reasons we can't have autoloaded functions yet. It is better if you explicitly call the root namespace function, either by doing `\php_sapi_name()`, or:
```php
use function php_sapi_name;
...
php_sapi_name();
```

The benefit of this is that it doesn't do the duplicated function lookup. This is a good habit to get into because there are certain functions that PHP can optimize as long as it knows you want the root namespaced version (see [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048)) for some discussion around this

oscarotero commented:
Ok, got it.
Would it be posible to move this detection to other method, for example `private function isCli(): bool`, in order to mock this method instead the global function?

duncan3dc commented:
That sounds fine, but I don't really have any more time to work on this now sorry. I can delete the uopz stuff and tests if you want to merge this pr and and refactor the code after?

oscarotero commented:
Ok, don't worry. I'll merge this as is, and then will make these changes.
Thanks for your work!

duncan3dc commented:
That's great, thanks for the project :+1:

oscarotero commented:
Mmm, one question.
Do you want use the default handler even if the request has the `text/html` value in the Accept header?

duncan3dc commented:
For my specific use case yes, as I want to basically replace the `PrettyPageHandler` with my handler.
But I guess it should be more flexible than that. Perhaps we should pass in a `HandlerFactory` that can receive the content type (or any other context) and provide the appropriate handler?

oscarotero commented:
Yes. That sounds to a PSR-11 container, so internally the middleware could do something like this:

```php
$accept = $request->getHeaderLine('Accept');
$handler = $this->container->get($accept);
```

duncan3dc commented:
Using a container would be a bit limiting, as you've only the generic string to pass to `get()`.
By creating a `HandlerFactory` we can support any other options that come up in the future, or just pass the whole request to it:
```php
$handler = $this->factory->getHandler($request);
```

oscarotero commented:
Yes, you're right but I prefer now to rely on PSR interfaces for a better interoperatibility.
I've commited the new [WhoopsHandlerContainer](https://github.com/middlewares/whoops/blob/master/src/WhoopsHandlerContainer.php). You could create a new container extending this and ovewrite the desired method (json(), html(), etc). Let me know if this covers your use case.

duncan3dc commented:
Yep that should work for my use case, thanks :+1:

duncan3dc commented:
Will you be tagging a new release for this soon?

oscarotero commented:
v1.1.0 released.
(I hope Whoops release a new version soon fixing the PHP 7.2 issues)
